module-system/sov-modules-api/src/lib.rs

```diff
@@ -301,46 +301,66 @@ pub trait RpcRunner {
-    modules: Vec<&dyn ModuleInfo<Context = C>>,
-) -> Result<Vec<&dyn ModuleInfo<Context = C>>, anyhow::Error> {
-    fn visit_module<'a, C: Context>(
+struct ModuleVisitor<'a, C: Context> {
+    visited: HashSet<&'a <C as Spec>::Address>,
+    sorted_modules: std::vec::Vec<&'a dyn ModuleInfo<Context = C>>,
+}
+
+impl<'a, C: Context> ModuleVisitor<'a, C> {
+    pub fn new() -> Self {
+        Self {
+            visited: HashSet::new(),
+            sorted_modules: Vec::new(),
+        }
+    }
+
+    fn visit_modules(
+        &mut self,
+        modules: Vec<&'a dyn ModuleInfo<Context = C>>,
+    ) -> Result<(), anyhow::Error> {
+        let mut module_map = HashMap::new();
+
+        for module in &modules {
+            module_map.insert(module.address(), *module);
+        }
+
+        for module in modules {
+            self.visit_module(module, &module_map)?;
+        }
+
+        Ok(())
+    }
+
+    fn visit_module(
+        &mut self,
         module: &'a dyn ModuleInfo<Context = C>,
-        visited: &mut HashSet<&'a C::Address>,
-        sorted_modules: &mut std::vec::Vec<&'a (dyn ModuleInfo<Context = C> + 'a)>,
-        module_map: &HashMap<&<C as Spec>::Address, &'a (dyn ModuleInfo<Context = C> + 'a)>,
+        module_map: &HashMap<&<C as Spec>::Address, &'a (dyn ModuleInfo<Context = C>)>,
     ) -> Result<(), anyhow::Error> {
         let address = module.address();
 
         // if the module hasn't been visited yet, visit it and its dependencies
-        if visited.insert(address) {
+        if self.visited.insert(address) {
             for dependency_address in module.dependencies() {
                 let dependency_module = *module_map.get(dependency_address).ok_or_else(|| {
                     anyhow::Error::msg(format!("Module not found: {:?}", dependency_address))
                 })?;
-                visit_module(dependency_module, visited, sorted_modules, module_map)?;
+                self.visit_module(dependency_module, module_map)?;
             }
 
-            sorted_modules.push(module);
+            self.sorted_modules.push(module);
         }
 
         Ok(())
     }
+}
 
-    let mut module_map = HashMap::new();
-
-    for module in &modules {
-        module_map.insert(module.address(), *module);
-    }
-
-    let mut visited = HashSet::new();
-    let mut sorted_modules = Vec::new();
-    for module in modules {
-        visit_module(module, &mut visited, &mut sorted_modules, &module_map)?;
-    }
-
-    Ok(sorted_modules)
+/// Sorts ModuleInfo objects by their dependencies
+pub fn sort_modules_by_dependencies<C: Context>(
+    modules: Vec<&dyn ModuleInfo<Context = C>>,
+) -> Result<Vec<&dyn ModuleInfo<Context = C>>, anyhow::Error> {
+    let mut module_visitor = ModuleVisitor::<C>::new();
+    module_visitor.visit_modules(modules)?;
+    Ok(module_visitor.sorted_modules)
 }
 
 /// Accepts Vec<> of tuples (&ModuleInfo, &TValue), and returns Vec<&TValue> sorted by mapped module dependencies
```

Comment:
It looks to me like we don't rely on the fact that `TValue` is passed by reference. Am I missing something? If not, we should use this signature: 
``rust
pub fn sort_values_by_modules_dependencies<C: Context, TValue>(
    module_value_tuples: Vec<(&dyn ModuleInfo<Context = C>, TValue)>,
) -> Result<Vec<TValue>, anyhow::Error>
```
That signature allows you to pass types by reference, but lets you pass by value as well if you want and removes  the lifetime constraint.

Also, I know I pushed for the rename from `T` which became `TValue` - but with the new signature I don't think we need it anymore. Now it's very clear that T is not related to the module in any way except for being part of the tuple - so `T` is idiomatic.